src/dhcpmon/src/dhcp_device.c

```diff
@@ -242,52 +235,122 @@ static void read_callback(int fd, short event, void *arg)
 }
 
 /**
- * @code dhcp_device_validate(counters, counters_snapshot);
+ * @code dhcp_device_is_dhcp_inactive();
+ *
+ * @brief Check if there were no DHCP activity
+ *
+ * @return 0 if there were no DHCP activity, -1 otherwise
+ */
+static int dhcp_device_is_dhcp_inactive()
+{
+    int rv = -1;
+
+    if ((aggregate_dev.counters[DHCP_RX].discover == aggregate_dev.counters_snapshot[DHCP_RX].discover) &&
+        (aggregate_dev.counters[DHCP_RX].offer    == aggregate_dev.counters_snapshot[DHCP_RX].offer   ) &&
+        (aggregate_dev.counters[DHCP_RX].request  == aggregate_dev.counters_snapshot[DHCP_RX].request ) &&
+        (aggregate_dev.counters[DHCP_RX].ack      == aggregate_dev.counters_snapshot[DHCP_RX].ack     )    ) {
+        rv = 0;
+    }
+
+    return rv;
+}
+
+/**
+ * @code dhcp_device_check_positive_health(counters, counters_snapshot);
+ *
+ * @brief validate current interface counters by comparing aggregate counters with snapshot counters.
+ *
+ * @param counters              recent interface counter
+ * @param counters_snapshot     snapshot counters
+ *
+ * @return DHCP_MON_STATUS_HEALTHY, DHCP_MON_STATUS_UNHEALTHY, or DHCP_MON_STATUS_INDETERMINATE
+ */
+static dhcp_mon_status_t dhcp_device_check_positive_health(dhcp_device_counters_t *counters,
+                                                           dhcp_device_counters_t *counters_snapshot)
+{
+    dhcp_mon_status_t rv = DHCP_MON_STATUS_HEALTHY;
+
+    // if we have rx DORA then we should have corresponding tx DORA (DORA being relayed)
+    if (((counters[DHCP_RX].discover >  counters_snapshot[DHCP_RX].discover) &&
+         (counters[DHCP_TX].discover <= counters_snapshot[DHCP_TX].discover)    ) ||
+        ((counters[DHCP_RX].offer    >  counters_snapshot[DHCP_RX].offer   ) &&
+         (counters[DHCP_TX].offer    <= counters_snapshot[DHCP_TX].offer   )    ) ||
+        ((counters[DHCP_RX].request  >  counters_snapshot[DHCP_RX].request ) &&
+         (counters[DHCP_TX].request  <= counters_snapshot[DHCP_TX].request )    ) ||
+        ((counters[DHCP_RX].ack      >  counters_snapshot[DHCP_RX].ack     ) &&
+         (counters[DHCP_TX].ack      <= counters_snapshot[DHCP_TX].ack     )    )    ) {
```

Comment:
Not that we would encounter this soon. A safe way to compare any integer that could wrap is like following:

bool less(unit32_t left, unit32_t right)
{
    return (int32_t)(left - right) < 0;
}

Reply:
A small style suggestion, move { to a new line when if condition extends multiple lines. Helps reading code.

Reply:
Yes, I am aware of this for modulu arithmetic as well. It is famous for sequence number of certain width (5, 7, 10 bits.) The counters here are 64 bit counters and are not expected to wrap around any time soon. They are reflecting actual counts of received packets since start of operation.

For style, I did code refactor that will make the code streamlined.